Add UserMute to control message visibility

grekko opened the pull request:
## Why are the changes necessary?

Implements the UserMute-Feature as described in https://github.com/osmfoundation/ewg_bidding/tree/main/projects/ability-to-mute-other-users/proposals/gregory_igelmund-08-09-2023

Addresses #3129


## What changed?

- Refactors `app/views/messages/*.html.erb` templates (extracting shared code)
- Adds `UserMute`-Model w/ CRUD operations to mute/unmute users
- Extends `Message`-Model with a `muted`-flag
- Adjusts `UserMailer.message_notification`-calls to respect the `Message#muted?`-flag

grekko commented on db/structure.sql:
```diff
@@ -3510,3 +3597,5 @@ INSERT INTO "schema_migrations" (version) VALUES
 ('7'),
 ('8'),
 ('9');
+
+
```
```suggestion
```

grekko commented on config/locales/en.yml:
```diff
@@ -2900,6 +2912,23 @@ en:
+        one: "You have muted %{count} User"
+        other: "You have muted %{count} Users"
+      user_mute_explainer: "Private messages of muted users are moved into a separate inbox and you won't receive email notifications."
+      user_mute_admins_and_moderators: "You can mute Admins and Moderators but their private messages cannot."
```
```suggestion
      user_mute_admins_and_moderators: "You can mute Admins and Moderators but their private messages cannot be muted."
```

tomhughes replied:
That change is of course equivalent to saying "you can mute them but it has no effect" so should we just not allow them to be muted?

I guess if their status changes later then the mute would suddenly activate but then that raises more questions...

grekko replied:
> That change is of course equivalent to saying "you can mute them but it has no effect" so should we just not allow them to be muted?

right. Users are able to mute Admins/Moderators but muting doesn't have an effect.

grekko replied:
> so should we just not allow them to be muted?

🤦🏽 I rushed over your text. You mean that we could just not display the `Mute this user`-action!

> I guess if their status changes later then the mute would suddenly activate but then that raises more questions...

Exactly this.

grekko commented on config/locales/en.yml:
```diff
@@ -2900,6 +2912,23 @@ en:
+      title: "Muted Users"
+    create:
+      notice: "You muted %{name}."
+      error: "Something went wrong."
+    destroy:
+      notice: "You unmuted %{name}."
+      error: "Something went wrong."
```
Either use controller action specific messages for the errors or reuse generic messages.

grekko commented:
Hey @gravitystorm 👋🏽

this is my take on implementing the `UserMute`-Feature. There are still things I currently consider to be missing and some things I would like to put up for discussion.

Before going into these I wanted to ask you to have a peek at the current shape of the code and let me know if you notice some issues with the fundamental structures though before I'd like into the details.

---

What I think is missing or I want to look more closely at:
- I am not super happy with the `before_save`-Callback in `Message` which calls out to `UserMute.active_for?` but haven't taken the time yet to think about a solid alternative that I'd prefer here.
- Add tests for new/changed controllers
- There are some system and integration tests. I might add an integration test since no JS is involved and the integration test would imo be appropriate here.

---

Open questions from the top of my head:

- I think the most natural place where users would expect the "UserMute"-action would be the `Messages#show`-screen, but I could not come up with a fitting UI-solution. Currently the "UserMute"-action lives just on the `Users#show`-screen. If it would live in `Messages#show`:
  - Where should this thing be placed? I'd expect it to be close to the Displayname/User image, not underneath the image, but I don't it to be rendered more prominent than necessary
  - Should it be a link "Mute this user" or a button?

- Users can move muted Messages from their `Muted messages`-Tab on `/messages/muted` back into the regular `My Inbox`-Tab. Should they be also able – at this point – to move Messages from the `My Inbox`-Tab to the `Muted messages`-Tag?

- Do we want to encourage users who have muted other users to remove the `UserMute`-again to prevent an accumulation of these mutes or should we wait how this feature plays out – and if it is even used? I am a fan of the latter.

tomhughes commented on app/models/user_mute.rb:
```diff
@@ -0,0 +1,45 @@
+#
+class UserMute < ApplicationRecord
+  belongs_to :creator, :class_name => "User"
+  belongs_to :appointee, :class_name => "User"
```
I think `appointee` is a pretty odd term to use for the muted user.

I mean I could argue for `mutor` and `mutee` for the creator and person being muted but that might be taking english a bit far.

Maybe `subject` for the person being muted?

On the other side `creator` isn't terrible but maybe `owner` is better? Not sure either really captures the concept well though.

grekko replied:
> Maybe `subject` for the person being muted?

I like `subject` much better!

I initially used `target` but didn't like it and went with `appointee` 😬 `subject` is better. I'll adjust it.

grekko replied:
> … maybe `owner` is better?

I went with `creator` b/c `creator` is used for `UserBlock`-records but `owner` is imho also expressive and being used e.g. in the concept of `OAuth`-associated records.

I am not having feelings strong enough towards 1 or the other and am happy to go with `owner`. I'll adjust this.

tomhughes commented on app/models/user_mute.rb:
```diff
@@ -0,0 +1,45 @@
+      )
+  end
+
+  def self.active_for?(current_user:, other_user:)
```
Why use different names for the users in the argument list here? It's the same concept as the member names so why not use the same names?

Of course once you do that the question becomes is this method even needed when it's just forwarding the arguments to `exists?`?

grekko replied:
> Of course once you do that the question becomes is this method even needed when it's just forwarding the arguments to exists??

good point. Initially there were more places from which `UserMute.active_for?` in which the the terms `current_user` and `other_user` are used to reference the instances. So I thought it was making the code a bit easier to read/grasp from that context.

Now there is just a single call, plus the call from `UserMute#for_message?` – I think it's ok to drop `UserMute#active_for?` and just call `UserMute#exists?` directly 👌🏽

tomhughes commented on db/migrate/20231010201451_create_user_mutes.rb:
```diff
@@ -0,0 +1,15 @@
+class CreateUserMutes < ActiveRecord::Migration[7.0]
+  def change
+    create_table :user_mutes do |t|
+      t.references :creator, :null => false
```
I think we can add `:index => false` here as we have a separate index on creator+appointee for uniqueness so also having one on creator is unnecessary.

tomhughes reviewed (commented):
Just a few initial comments, not a detailed review.

grekko commented:
@gravitystorm ping. Could you have a look at the general direction of the code in this PR? Do you have already some notes for me?

tomhughes commented on app/controllers/user_mutes_controller.rb:
```diff
@@ -0,0 +1,46 @@
+      flash[:error] = t(".error")
+    end
+
+    redirect_back :fallback_location => user_mutes_path(current_user)
```
It seems `redirect_back_or_to` is now preferred over `redirect_back` (that might be new in rails 7.1) though both work for now.

tomhughes commented on app/views/user_mutes/index.html.erb:
```diff
@@ -0,0 +1,36 @@
+          </td>
+          <td class="text-nowrap text-end">
+            <%= link_to t(".unmute"), unmute_user_path(user), :method => :delete, :class => "btn btn-sm btn-primary" %>
+            <%= link_to t(".send_message"), new_message_path(user), :class => "btn btn-sm btn-secondary" %>
```
An option to send a message to a muted user seems a bit strange when by definition any reply will wind up in the muted messages and won't be notified?

grekko replied:
> An option to send a message to a muted user seems a bit strange when by definition any reply will wind up in the muted messages and won't be notified?

I can imagine situations in which I muted another user just to prevent them from spamming my email inbox but still want to keep the conversation open.

tomhughes commented on config/locales/en.yml:
```diff
@@ -1724,9 +1729,16 @@ en:
+    heading: 
+      my_inbox: "My Inbox"
+      my_outbox: "My Outbox"
+      muted_messages: "Muted messages"
```
Should the capitalisation be consistent here given these are three labels that sit adjacent to each other?

grekko replied:
To me the 3 label terms here are different. My reasoning:

- `Inbox` and `Outbox` are defined terms and therefore should be capitalised.
- `Muted messages` is not.

Bear in mind that I am not a native english speaker so I'm happy to adjust the label to `Muted Messages if you say so.

tomhughes commented on config/locales/en.yml:
```diff
@@ -2900,6 +2915,23 @@ en:
+      you_have_muted_n_users:
+        zero: "You haven't muted any Users 🎉"
+        one: "You have muted %{count} User"
+        other: "You have muted %{count} Users"
```
Why are we capitalising User/Users here?

grekko replied:
Because all actions containing the keyword `User` on `Users#show` are currently capitalised.

![SCR-20231023-lnud](https://github.com/openstreetmap/openstreetmap-website/assets/28648/769b4874-e74c-4896-8dbc-355ef5c2a8f5)

grekko replied:
ah. I understand that here in this context "Users" should not be capitalised since it's not referencing a language subject. I'll adjust the translation.

tomhughes reviewed (changes requested):
In addition to the specific comments I've made, a few general observations/questions...

Firstly I think we should have a "mute user" button on the detail view (messages#show) for messages to make it more obvious that the option is available when a problem arises? Probably we should also have report user (or ideally report message) but that's not really in scope for this project.

Secondly should we have a mute/unmute button on the main message view - currently messages can only be unmuted from the index page and there's no way to mute a message at all if the user is not muted when it is received.

gravitystorm commented on app/controllers/messages_controller.rb:
```diff
@@ -127,6 +134,23 @@ def mark
 
+  # Moves message into Inbox by unsetting the muted-flag
+  def unmute
+    message = Message.where(:recipient => current_user).find(params[:message_id])
```
We prefer `find_by` instead of `where().first`, but `Model.find(id)` is still preferred to `Model.find_by(:id => id)`.

I think some kind of scoping is necessary to avoid unmuting other people's messages, although I'd normally try to find a `current_user.something.find()` to scope them rather than building a where query.

grekko replied:
> This never actually changes the state of the message, so nothing happens when you click the Unmute button for a message in the user interface.

this is just the query for the `message`-record. 2 lines below `message.unmute` is called.

grekko replied:
> I think some kind of scoping is necessary to avoid unmuting other people's messages

The query is already scoped to the `current_user`

```rb
Message.where(:recipient => current_user).find(params[:message_id])
```

But yes, I prefer a scope on the `current_user`, too. I'll change this too:

```rb
current_user.muted_messages.find(params[:message_id])
```

tomhughes replied:
> this is just the query for the `message`-record. 2 lines below `message.unmute` is called.

Oh I missed that... In that case the question becomes why wasn't it working for me because the reason I stared looking closely at that code was that the unmute didn't seem to work.

grekko replied:
> In that case the question becomes why wasn't it working for me

not sure what went wrong there :)

grekko commented:
Regarding the general feedback from @tomhughes 

> Firstly I think we should have a "mute user" button on the detail view (messages#show) for messages to make it more obvious that the option is available when a problem arises? […]

I think so too, but I haven't found the ideal place yet where I want to put that `mute user`-action.

This is the `Messages#show`-screen:

![SCR-20231023-mdgi](https://github.com/openstreetmap/openstreetmap-website/assets/28648/ae3617fb-9702-4f5c-b140-6aeeab8e31a9)

The action should be imo close to the Username (at the top) and not at the bottom since all actions on the bottom of the message are message-related actions.

There is not much space at the top of the message. We could re-use the layout we already use on `Dashboards#show`.

![SCR-20231023-njqe](https://github.com/openstreetmap/openstreetmap-website/assets/28648/7911199e-f6ef-4f5c-a37d-b60fa1de1d85)


The updated `Messages#show`-screen could look like this then:

![SCR-20231023-nivc](https://github.com/openstreetmap/openstreetmap-website/assets/28648/70939f54-dfd6-4df8-a990-1c6afc30c2f3)

grekko commented:
> Secondly should we have a mute/unmute button on the main message view - currently messages can only be unmuted from the index page and there's no way to mute a message at all if the user is not muted when it is received.

I understand your point. I imagine though that it would add unnecessary noise to the UI for many users when we add a `Mute` or `Move to muted Messages`-button next to each `Inbox`-message:

![SCR-20231023-nnbv](https://github.com/openstreetmap/openstreetmap-website/assets/28648/d8251b11-2e53-4c77-9533-b0fb53ae23cc)

vs.

![SCR-20231023-nnyk](https://github.com/openstreetmap/openstreetmap-website/assets/28648/7567f651-2edb-4c85-a6d5-14b591802223)

In my understanding only a fraction of users – lets say 5% for the sake of my argument – will have to use the `UserMute`-Feature on rare occasions.

Adding the `Mute`-action to all Inbox-messages for all users makes the `UserMute`-Feature a bit too prominent since 95% of the users never have to bother with the Feature.

What do you think? @tomhughes

gravitystorm reviewed (commented):
I'm sorry I haven't had the opportunity to do a full review - this is my initial thoughts from reading the code, without trying out the interfaces. Happy to discuss any of these points.

grekko commented on app/controllers/messages_controller.rb:
```diff
@@ -47,7 +47,7 @@ def create
     elsif @message.save
       flash[:notice] = t ".message_sent"
-      UserMailer.message_notification(@message).deliver_later
+      @message.notify_recipient_via_mail
```
> I think it should be the controller that decides what is the list of things that should be done.

I think that is still the case. The controller still makes the call, but now there is this extra logic which is imo a detail that the controller does not need to bother about.

I understand though that currently all direct calls to the `UserMailer` are made either from within the context of a Controller or Job and my change deviates from that pattern.

To me it is important that the logic wether a E-Mail Notification should be sent is kept in a single place and is not spread across the application code.

What do you think about this version?

```diff
diff --git a/app/controllers/messages_controller.rb b/app/controllers/messages_controller.rb
index 21d08e7af..2ca86fc02 100644
--- a/app/controllers/messages_controller.rb
+++ b/app/controllers/messages_controller.rb
@@ -47,7 +47,7 @@ class MessagesController < ApplicationController
       render :action => "new"
     elsif @message.save
       flash[:notice] = t ".message_sent"
-      @message.notify_recipient_via_mail
+      UserMailer.message_notification(@message).deliver_later if @message.notify_recipient?
       redirect_to :action => :inbox
     else
       @title = t "messages.new.title"
diff --git a/app/models/message.rb b/app/models/message.rb
index 1be482cfb..fdeb89c4b 100644
--- a/app/models/message.rb
+++ b/app/models/message.rb
@@ -74,10 +74,8 @@ class Message < ApplicationRecord
     md5.hexdigest
   end
 
-  def notify_recipient_via_mail
-    return false if muted?
-
-    UserMailer.message_notification(self).deliver_later
+  def notify_recipient?
+    !muted?
   end
 
   def unmute
diff --git a/script/deliver-message b/script/deliver-message
index a382bffa1..496773597 100755
--- a/script/deliver-message
+++ b/script/deliver-message
@@ -33,6 +33,6 @@ mail = Mail.new($stdin.read
 message = Message.from_mail(mail, from, to)
 message.save!
 
-message.notify_recipient_via_mail
+UserMailer.message_notification(message).deliver_later if message.notify_recipient?
 
 exit 0
```

grekko replied:
I'll apply the patch and will ask you for a re-review.

grekko replied:
The patch was applied => Closing the conversation (for now).

grekko commented on app/controllers/messages_controller.rb:
```diff
@@ -107,6 +107,13 @@ def outbox
+  def muted
+    @title = t ".title"
+
+    redirect_to inbox_messages_path if current_user.muted_messages.none?
```
The redirect is meant to handle the situation in which a user moves the last muted message back into the Inbox, like so:

![2023-10-26 20 40 42](https://github.com/openstreetmap/openstreetmap-website/assets/28648/16ca6c81-bca2-4f98-ac55-8f6033eed7f4)

Once there are no muted Messages anymore, the `Muted`-Tab is hidden and the user is redirected back to the Inbox.

Rendering the template instead would create an inconsistent UI because no Tab would be active. Or we would need to display the `Muted`-Tab even though there are no messages and then hide it once the user clicks on any other link.

Why would you prefer rendering the template?

grekko replied:
We also hide the `Muted users`-Tab on the Account-page, so I understand that we settled on this behaviour (for now) => resolving the conversation.

grekko commented on app/models/user_mute.rb:
```diff
@@ -0,0 +1,36 @@
+
+  validates :subject, :uniqueness => { :scope => :owner_id, :message => :is_already_muted }
+
+  def self.for_message?(message)
```
> […] the message object isn't being used for anything here, other than a sort of container for the two users.

To me having the method `.for_message?` expresses that the `UserMute` has this special relation to messages. When reading this I'd think that `UserMute` is designed to be working only in conjunction with messages. I'm aware that this is my subjective interpretation though and I not strongly for this specific implementation.

You were thinking about something like this?

```diff
diff --git a/app/models/message.rb b/app/models/message.rb
index 1be482cfb..bfb9c6a75 100644
--- a/app/models/message.rb
+++ b/app/models/message.rb
@@ -87,6 +87,6 @@ class Message < ApplicationRecord
   private
 
   def set_muted
-    self.muted ||= UserMute.for_message?(self)
+    self.muted ||= UserMute.active?(owner: recipient, subject: sender)
   end
 end
diff --git a/app/models/user_mute.rb b/app/models/user_mute.rb
index e0b80d77a..311ca0bf7 100644
--- a/app/models/user_mute.rb
+++ b/app/models/user_mute.rb
@@ -23,13 +23,11 @@ class UserMute < ApplicationRecord
 
   validates :subject, :uniqueness => { :scope => :owner_id, :message => :is_already_muted }
 
-  def self.for_message?(message)
-    sender = message.sender
-
+  def self.active?(owner:, subject:)
     !sender.administrator? &&
       !sender.moderator? &&
       exists?(
-        :owner => message.recipient,
+        :owner => owner,
         :subject => sender
       )
   end
```

grekko replied:
I'll apply the patch and will ask you for a re-review.

grekko replied:
The patch was applied => Closing the conversation (for now).

grekko commented on test/models/user_mute_test.rb:
```diff
@@ -0,0 +1,27 @@
+      create(:user_mute, :owner => user, :subject => admin_or_moderator)
+
+      message = build(:message, :sender => admin_or_moderator, :recipient => user)
+      message.save
```
🤔 yeah, I would expect so. I am not sure the code calls `build` and `save` – I think this can and should just be `create` – thanks!

grekko replied:
I updated the original commit and force pushed the changes. => resolving this conversation.

grekko commented on app/abilities/ability.rb:
```diff
@@ -47,13 +47,14 @@ def initialize(user)
         can [:new, :create, :edit, :update, :comment, :subscribe, :unsubscribe], DiaryEntry
         can [:make_friend, :remove_friend], Friendship
-        can [:new, :create, :reply, :show, :inbox, :outbox, :mark, :destroy], Message
+        can [:new, :create, :reply, :show, :inbox, :outbox, :muted, :mark, :unmute, :destroy], Message
```
Thanks for bringing this up. I read the article and the described patterns and ideas sounds reasonable at first glance but I have many open questions on how I'd implement specific current non-CRUD operations in a CRUD-style.

Splitting up the `Messages#inbox`, `Messages#outbox` and `Messages#muted` actions into separate controllers, e.g. the `InboxController#index` you already mentioned seems straight forward to me.

Although I'd probably namespace those controllers e.g. like to `Messages::InboxController#index` just to make it very clear that this is about a Message Inbox and not any other Inbox.

Also I would say that `Messages#reply` could be `Messages::Replies#create` instead – in the spirit of keeping things CRUD-oriented – although I could be reading this as "This controller action creates a `Message::Reply`-record, although it actually creates just another `Message`-record 🤔 

---

The ideas of keeping things more uniform and simple resonate with me, but I'd need more time to fiddle around with the ideas and how to apply them without creating "just more code and files" without getting the benefit of more organised and easier to grasp code and files.

I must admit: Thats something I would not feel comfortable with experimenting in the scope of this PR :)

gravitystorm replied:
> I must admit: Thats something I would not feel comfortable with experimenting in the scope of this PR :)

Sure, no worries! I just thought I would ask.

tordans commented:
@grekko Does this already handle the API part? We have this part here https://wiki.openstreetmap.org/wiki/API_v0.6#Details_of_the_logged-in_user:_GET_/api/0.6/user/details

I think, the easiest would be to change the queries for `<received count="1" unread="0"/>` to only include messages by unmuted users. However, I wonder if we should also add the full count just to give the API access to all the parts that the UI has as well. Something like `<received count="1" unread="0"/><received_mute count="1" unread="0"/>`

I would both not consider this a breaking change of the API. Changing the way the data is caculated is in line with how the new UI behaves. And adding something does not break (a well build api consumer).

(Disclaimer: I only did a quick scan of the files-list to check if this was solved already, sorry if it was…)

grekko commented:
> Does this already handle the API part? We have this part here

The code for the user serialisation is here https://github.com/openstreetmap/openstreetmap-website/blob/0844024c8e68ae0031e143d6094793f0aa32bc22/app/views/api/users/_user.json.jbuilder#L59-L63

The `muted` Messages are excluded here and not mentioned explicitly, so right know it behaves identical to the UI.

If API consumers want/need access to the `muted` Message count, one could add it the the serialised user object, but I didn't add it without having a use case here.

grekko commented:
Hey @gravitystorm and/or @tomhughes 👋🏽 

Any chance to get a PR review with comments or Follow-Ups on the open PR conversations?
I would really like to get the PR in a mergeable shape.

gravitystorm commented on app/views/messages/_messages_table.html.erb:
```diff
@@ -0,0 +1,13 @@
+    </tr>
+  </thead>
+  <tbody>
+    <%= render :partial => inner_partial, :collection => messages %>
```
There is a level of indirection here with isn't used, since "inner_partial" is always equal to "message_summary". I guess what might have happened is that you started looking at the two different contents for this table - inbox and outbox being slightly different - but you haven't actually refactored the outbox, perhaps because the headings are different too.

So I think either more refactoring is required (so that the inner_partial is useful) or this level of indirection can be removed.

grekko replied:
see c0cbf1c28da68739bc854eb8249f5298ce16e8a5

gravitystorm commented on config/locales/en.yml:
```diff
@@ -2906,6 +2927,30 @@ en:
+    index:
+      title: "Muted Users"
+      my_muted_users: "My muted users"
+      you_have_muted_no_users: "You have not muted any users 🎉"
```
No celebration emoji here please! I'm no big fan of emojis (they are an i18n nightmare), but this one should be used when something good happens to you, or you've taken a successful action, so it doesn't really fit the context of "nothing bad happened yet".

gravitystorm commented on config/locales/en.yml:
```diff
@@ -2906,6 +2927,30 @@ en:
+      you_have_muted_n_users:
+        one: "You have muted %{count} User"
+        other: "You have muted %{count} users"
+      user_mute_explainer: "Private messages of muted users are moved into a separate inbox and you won't receive email notifications."
```
We don't use the term "private messages" anywhere else in the UI, only "messages". So this could be confusing if someone is trying to work out if "private messages" are different from regular messages.

I think "private" can simply be removed in these two translations.

gravitystorm commented on config/routes.rb:
```diff
@@ -289,6 +292,11 @@
 
+  # muting users
+  post "/user/:display_name/mute" => "user_mutes#create", :as => :mute_user
+  delete "/user/:display_name/mute" => "user_mutes#destroy", :as => :unmute_user
```
If these can be added as resourceful routes then that would be better. Or at lest, specific actions on a resource or scoped resource. Something like (untested)

```ruby
scope "/user/:display_name" do
  resources :user_mutes, :only -> [:create, :destroy]
end
```

We're (slowly) getting rid of all the "full handwritten" routes so I'm keen to avoid adding more.

gravitystorm commented on test/integration/user_muting_test.rb:
```diff
@@ -0,0 +1,31 @@
+require "test_helper"
+
+class UserMutingTest < ActionDispatch::IntegrationTest
```
This should perhaps be a system test rather than an integration test, where we can use more natural things like "click_button" rather than `assert_select` followed by directly posting to controllers.

gravitystorm commented on test/integration/user_muting_test.rb:
```diff
@@ -0,0 +1,31 @@
+    get inbox_messages_path
+    assert_select "a[href='#{muted_messages_path}']", false
+
+    session_for(other_user)
```
I think just `create(:message, ....)` rather than switching sessions here

grekko replied:
I've transformed the integration test into a system spec and split up the test cases so that no session switch/or sign_in/sign_out is necessary anymore to verify the behaviour.

gravitystorm commented on test/models/user_mute_test.rb:
```diff
@@ -0,0 +1,26 @@
+    create(:user_mute, :owner => user, :subject => muted_user)
+
+    message = build(:message, :sender => muted_user, :recipient => user)
+    message.save
```
`create` rather than `build` followed immediately by `save`?

gravitystorm commented on app/views/application/_settings_menu.html.erb:
```diff
@@ -14,5 +14,8 @@
       <%= link_to t(".oauth2_authorizations"), oauth_authorized_applications_path, :class => "nav-link #{'active' if controller_name == 'oauth2_authorized_applications'}" %>
     </li>
+    <li class="nav-item">
+      <%= link_to t(".muted_users"), user_mutes_path, :class => "nav-link #{'active' if controller_name == 'user_mutes'}" %>
```
We hide the "Muted messages" tab when there's no muted messages. Should we also hide the "Muted users" tab when you haven't muted anyone yet? There's not much value for this tab until you do (and the muting happens elsewhere), and most people are (hopefully) never going to need it.

grekko replied:
yup. I changed `messages/outbox.html.erb` to also reuse `messages/_messages_table.html.erb` and therefore pass a custom `inner_partial` and a new configuration – `columns` to `messages/_messages_table.html.erb`.

I'd understand if the view logic seems a bit overcomplicated now, but I think its still not too hard to read and therefore the message table logic is kept together in a small set of files.

See c0cbf1c28da68739bc854eb8249f5298ce16e8a5

gravitystorm reviewed (changes requested):
In general this looks good to me, just a few more notes on things I would like to see before it's ready to go.

grekko commented:
@gravitystorm thanks for the review comments! I think I've addressed them all.

I just noticed and addressed 1 bug and 1 odd thing:
- The `Mute this User`-action on `app/views/users/show.html.erb` was displayed to unauthenticated users. I think this was confusion.
- The redirects in `UserMutes#destroy` after Unmuting the last muted user caused a redirect loop – not anymore.

grekko commented on config/locales/en.yml:
```diff
@@ -1726,9 +1735,16 @@ en:
       as_read: "Message marked as read"
       as_unread: "Message marked as unread"
+    unmute:
+      notice: "Message has been moved to inbox"
```
```suggestion
      notice: "Message has been moved to Inbox"
```

grekko commented on config/locales/en.yml:
```diff
@@ -1726,9 +1735,16 @@ en:
       as_unread: "Message marked as unread"
+    unmute:
+      notice: "Message has been moved to inbox"
+      error: "The message could not be moved to the inbox."
```
```suggestion
      error: "The message could not be moved to the Inbox."
```

grekko commented on script/deliver-message:
```diff
@@ -33,6 +33,6 @@ mail = Mail.new($stdin.read
 message.save!
 
-UserMailer.message_notification(message).deliver
+UserMailer.message_notification(message).deliver_later if message.notify_recipient?
```
```suggestion
UserMailer.message_notification(message).deliver if message.notify_recipient?
```

grekko commented:
I restored an unintended change (62fb859) and adjusted the locales so that "inbox" is written as "Inbox".

@gravitystorm Let me know if you spot anything else that you think needs adjustment.

gravitystorm commented:
> @gravitystorm Let me know if you spot anything else that you think needs adjustment.

The most obvious blocker at the moment is the git history, it would be great to clean that up and force push to this branch to update this PR.

I'm keen to get this merged, it looks pretty much ready to go now.

grekko commented:
> The most obvious blocker at the moment is the git history

Yup. I planned to invest the time to cleanup the git history once we feel comfortable to move forward. I will do this now then.

grekko commented:
@gravitystorm A fresh git history. I tried to cut it into meaningful parts. Let me know if you spot changes that you rather have merged into a single commit or split up.

gravitystorm reviewed (approved):
Thanks, I'm happy with this PR as-is, and I think it's ready to merge.

There were a couple of UI suggestions from @tomhughes that are not fully resolved in the discussion, and I would be happy to see them addressed as followups since I agree they are slightly tricky, and also I think they aren't blockers for the initial functionality.

grekko commented:
>  I'm happy with this PR as-is, and I think it's ready to merge.

Thanks a lot for the support and steering the process. 🙇🏽 

> There were a couple of UI suggestions…

I am happy to contribute changes in that direction. I see the changes are also rolled out and am currently testing them manually in production.

I am looking forward to see if the features get any traction and wether it adds the expected value for some users in the community.
